test/base/map.js

```diff
@@ -634,6 +634,18 @@ test("#1980 .replace should not breaks entities order!", () => {
     }
 })
 
+test("#1980 .replace should should invoke autorun", () => {
+    const original = mobx.observable.map({ a: "a", b: "b" })
+    const replacement = { b: "b", a: "a" }
+    let numOfInvokes = 0
+    autorun(() => {
+        numOfInvokes = numOfInvokes + 1
+        return original.entries()
+    })
+    original.replace(replacement)
+    expect(numOfInvokes).toBe(1)
```

Comment:
You have to expect 2 - autorun always runs at least once - during definition and later on every notification

Reply:
So chances are it actually doesn't. Good thing we did add that test.

Btw, I would also add expect inside the autorun to check for correct order, just to be sure.

Reply:
Alternatively you can use `reaction(() => map.entries(), () => numOfInvokes++)` and expect `1`

Reply:
I checked this issue, you're right but at the previous implementation, it failed too.
About the solution in the mobx new version, i can use **this._keysAtom.reportChanged()** but at mobx4 i dont have **k_eysAtom**.
Do you have any suggestions?

Reply:
I will try to take a look at v4 later this weekend or the next week. I think we have to modify `this._keys` or something ... not even sure why the impl is different (as there is no need for proxies?)

Reply:
@haimrait awesome work so far!

In addition, we should probably test that there are _no_ changes in the keys, the keyset is not updated (e.g. `{ a: "a", b: "b"}` -> `{ a: "a", b: "b" }` should not result in new keys nor entries, and `{ a: "a", b: "b" }` -> ` { a: "A", b: "B" }` should result in new entries but no new keyset)

If we get this correct, getting it right in v4 shouldn't be too hard indeed. (I'd actually expect the implementation to be the same indeed :P)

Reply:
@mweststrate you mean that we should check and not constantly assign
this._data = orderedData.
Because in { a: "a", b: "b"} -> { a: "a", b: "b" } case we have no change.
and { a: "a", b: "b" } -> { a: "A", b: "B" } case we want to change only keys value and not create a new Map object.
Can you confirm my statements?

Reply:
Hi @haimrait the constant reassignment is fine. I think the only thing needed to pass the test cases described above is to only trigger the `keysAtom.reportChanged` if there is actually a change in the keyset (including changes in order). So that should be conditional.

Reply:
but if we are doing the actual reassignment its change the this._data - Map entries and keys at the memory because it's a different map?

Reply:
First of all I think that the V4 could be refactored into V5 (`this._keys` => `this._keysAtom`) (?)
For the time being, if I follow the V4 code correctly, the `this._data` can stay in whatever order, but we have to make sure that `this._keys` is ordered correctly:
```javascript
replace(values: ObservableMap<K, V> | IKeyValueMap<V> | any): ObservableMap<K, V> {
    transaction(() => {
      const replacementMap = ObservableMap.convertToMap(values);      
      const oldKeys = this._keys;      
      const newKeys = Array.from(replacementMap.keys());      
      let keysChanged = false;
      for (let i = 0; i < oldKeys.length; i++) {
        const oldKey = oldKeys[i];
        // order key change        
        if (oldKeys.length === newKeys.length && oldKey !== newKeys[i]) {
          keysChanged = true;
        }
        // deleted key
        if (!replacementMap.has(oldKey)) {          
          keysChanged = true;
          this.delete(oldKey);
        }
      }
      replacementMap.forEach((value, key) => {
        // new key
        if (!this._data.has(key)) {
          keysChanged = true;
        }
        this.set(key, value);                        
      })      
      if (keysChanged) {
        this._keys.replace(newKeys);
      }      
    })
    return this
}
```

Reply:
@haimrait if you get stuck / confused somewhere, just let us know! I'll gladly help, and admittedly not following up very closely, as I'm currently in the middle of a move. But hope to have some more spare time again in 2 weeks or so!

Reply:
yea im sorry for the late i hope to commit changes today.